### ISSUE TITLE
Redefine dataclass variables with more concrete types in __post_init__ not recognised

### ISSUE THREAD

**Issue by airallergy** [python/mypy#13000]
**Bug Report**

After converting a dataclass variable to a more concrete type in `__post_init__`, mypy does not recognise the converted type in other methods.

**To Reproduce**

```python
from dataclasses import dataclass
from collections.abc import Sequence


@dataclass
class A:
    x: Sequence[int]

    def __post_init__(self):
        self.x = list(self.x)

    def foo(self) -> None:
        self.x.sort()
```

**Expected Behavior**

No error is expected, as redefining a variable with a more concrete type is allowed.

**Actual Behavior**

Mypy complains about the type. Specific mypy error codes depend on usage, e.g. `no attribute` when calling the converted type's method, or `incompatible type` when calling functions requiring the converted type.

**Your Environment**

- Mypy version used: 0.961
- Python version used: 3.10.5
- Operating system and version: macOS 12.4

**Comment by hauntsaninja** [comment]
+1 to Jelle, mypy should not be narrowing away from your declared type. dataclasses are mutable!

(Fwiw, I personally find wanting the equivalent of a non-trivial `__init__` in a dataclass to be a bit of an antipattern. Sometimes classmethod constructors and the default dataclass `__init__` are a good alternative. Sometimes just use a regular class)
